go/types/value_encoder.go

```diff
@@ -183,24 +191,24 @@ func (w *valueEncoder) writeCycle(i uint32) {
 	w.writeUint32(i)
 }
 
-func (w *valueEncoder) writeStructType(t *Type, parentStructTypes []*Type) {
-	// The runtime representaion of struct types can contain cycles. These cycles are broken when encoding and decoding using special "back ref" placeholders.
-	i, found := indexOfType(t, parentStructTypes)
-	if found {
-		w.writeCycle(uint32(len(parentStructTypes)) - i - 1)
+func (w *valueEncoder) writeStructType(t *Type, seenStructs map[string]*Type) {
+	desc := t.Desc.(StructDesc)
+	name := desc.Name
+
+	if _, ok := seenStructs[name]; ok {
+		w.writeKind(CycleKind)
+		w.writeString(name)
 		return
 	}
-	parentStructTypes = append(parentStructTypes, t)
+	seenStructs[name] = t
 
 	w.writeKind(StructKind)
-	desc := t.Desc.(StructDesc)
 	w.writeString(desc.Name)
-
 	w.writeUint32(uint32(desc.Len()))
 
 	for _, field := range desc.fields {
 		w.writeString(field.Name)
```

Comment:
yeah, let's go ahead and do that now just to keep consistency. i think the same argument about compression efficiency *might* apply in some cases.